### ISSUE TITLE
CS4031 should be reported for async lambda in [SecurityCritical] type

### ISSUE THREAD

**Issue by cston** [dotnet/roslyn#54074]
Compile the following:
```csharp
using System;
using System.Security;
using System.Threading.Tasks;

[SecurityCritical]
class Program
{
    static void Main()
    {
        Func<Task> f = async () => await Task.Yield();
    }
}
```
Expected:
```
error CS4031: Async methods are not allowed in an Interface, Class, or Structure
    which has the 'SecurityCritical' or 'SecuritySafeCritical' attribute.
```
Actual: No errors

**Comment by Youssef1313** [comment]
Small note: I think the message needs to be updated. In the sample above, the class `Program` itself doesn't contain any async methods (it contains a non-async method that contains an async lambda) ,but the message will be saying it does.

Probably update it to something like:

```
An interface, class, or struct with 'SecurityCritical' or 'SecuritySafeCritical' attribute is not allowed to contain async methods or methods with async lambdas.
```

**Comment by Youssef1313** [comment]
Also note that from the [attribute docs](https://docs.microsoft.com/en-us/dotnet/api/system.security.securitysafecriticalattribute?view=net-5.0):

> Partially trusted code is no longer supported. This attribute has no effect in .NET Core.

**Comment by jaredpar** [comment]
Don't think this is a high-pri item, and possibly "Won't Fix". As @Youssef1313 points out the scenario this impacts, Partial Trust, is no longer supported.
